tests/compiletest.rs

```diff
@@ -1,71 +1,432 @@
+                        eprint!("{} .. ", path.display());
+                        if errors.is_empty() {
+                            eprintln!("{}", "ok".green());
+                        } else {
+                            eprint!("{}", "FAILED".red().bold());
+                            if !revision.is_empty() {
+                                eprint!(" (revision `{}`)", revision);
+                            }
+                            eprintln!();
+                            failures.lock().unwrap().push((path.clone(), m, revision, errors));
+                        }
+                    }
+                }
+            });
+        }
+    })
+    .unwrap();
+
+    let failures = failures.into_inner().unwrap();
+    let total = total.load(Ordering::Relaxed);
+    let skipped = skipped.load(Ordering::Relaxed);
+    if !failures.is_empty() {
+        for (path, miri, revision, errors) in &failures {
+            eprintln!();
+            eprint!("{} {}", path.display().to_string().underline(), "FAILED".red());
+            if !revision.is_empty() {
+                eprint!(" (revision `{}`) ", revision);
+            }
+            eprintln!();
+            eprintln!("command: {:?}", miri);
+            for error in errors {
+                match error {
+                    Error::ExitStatus(mode, exit_status) => eprintln!("{mode:?} got {exit_status}"),
+                    Error::PatternNotFound(pat) =>
+                        eprintln!("error-pattern `{pat}` not found in stderr output"),
+                    Error::InlinePatternNotFound(pat) =>
+                        eprintln!("`{pat}` not found in stderr output"),
+                    Error::NoPatternsFound => eprintln!("no error patterns found in failure test"),
+                    Error::OutputDiffers { path, actual, expected } =>
+                        compare_output(path, actual, expected),
+                }
+            }
+        }
+        eprintln!();
+        eprintln!(
+            "{} tests failed, {} tests passed, {} skipped",
+            failures.len().to_string().red().bold(),
+            (total - failures.len() - skipped).to_string().green(),
+            skipped.to_string().yellow()
+        );
+        std::process::exit(1);
+    }
+    eprintln!();
     eprintln!(
-        "{}",
-        format!("## Running compile-fail tests in {} against miri for target {}", path, target)
-            .green()
-            .bold()
+        "{} tests passed, {} skipped",
+        (total - skipped).to_string().green(),
+        skipped.to_string().yellow()
     );
+}
+
+#[derive(Debug)]
+enum Error {
+    /// Got an invalid exit status for the given mode.
+    ExitStatus(Mode, ExitStatus),
+    PatternNotFound(String),
+    InlinePatternNotFound(String),
+    NoPatternsFound,
+    OutputDiffers {
+        path: PathBuf,
+        actual: String,
+        expected: String,
+    },
+}
+
+type Errors = Vec<Error>;
+
+fn revisions(path: &Path) -> Vec<String> {
```

Comment:
So we have a mini test runner lib here now. :D
Seems fine, I guess, but could this be a separate file from the "driver" logic?

Reply:
I want to make it a separate crate entirely

Reply:
Sure, though short-term it's nice to have this in-tree. But making it a separate file would help with reviewing so it would make sense to do that before landing IMO.

Reply:
Oh, I did mean a separate crate in tree. Just so I can write unit tests for the test runner itself